File: src/main/java/myLang/expression/EqualityExpressionContext.java
package myLang.expression;

import myLang.ContextHandler;
import myLang.response.BooleanResponse;
import myLang.response.MyLangException;
import myLang.response.Response;
import myLangParser.MyLangLexer;
import myLangParser.MyLangParser;
import org.antlr.v4.runtime.tree.TerminalNodeImpl;

public class EqualityExpressionContext implements ContextHandler<MyLangParser.EqualityExpressionContext> {

    @Override
    public Response handler(MyLangParser.EqualityExpressionContext ctx) throws MyLangException {
        if(ctx.getChildCount() == 1){
            return defaultHandler(ctx);
        }
        Response firstResponse = defaultHandler(ctx, 0);
        Response secondResponse = defaultHandler(ctx, 2);
        switch (((TerminalNodeImpl) ctx.getChild(1)).getSymbol().getType()) {
            case MyLangLexer.EQUAL:
                return new BooleanResponse(firstResponse.getResponse().equals(secondResponse.getResponse()));
            case MyLangLexer.NOTEQUAL:
                return new BooleanResponse(!firstResponse.getResponse().equals(secondResponse.getResponse()));
        }

        throw new MyLangException("Неподдерживаемая операция сравнения");
    }
}

File: src/main/kotlin/myLangKotlin/expression/EqualityExpressionContext.kt
package myLangKotlin.expression

import myLangKotlin.ContextHandler
import myLangKotlin.response.BooleanResponse
import myLangKotlin.response.MyLangException
import myLangKotlin.response.Response
import myLangParser.MyLangLexer
import myLangParser.MyLangParser
import org.antlr.v4.runtime.tree.TerminalNodeImpl

class EqualityExpressionContext : ContextHandler<MyLangParser.EqualityExpressionContext> {

    @Throws(MyLangException::class)
    override fun handler(ctx: MyLangParser.EqualityExpressionContext): Response<*> {
        if (ctx.childCount == 1) {
            return defaultHandler(ctx)
        }
        val firstResponse = defaultHandler(ctx, 0)
        val secondResponse = defaultHandler(ctx, 2)
        when ((ctx.getChild(1) as TerminalNodeImpl).getSymbol().type) {
            MyLangLexer.EQUAL -> return BooleanResponse(firstResponse.response == secondResponse.response)
            MyLangLexer.NOTEQUAL -> return BooleanResponse(firstResponse.response != secondResponse.response)
        }

        throw MyLangException("Неподдерживаемая операция сравнения")
    }
}
